nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py

```diff
@@ -0,0 +1,521 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+from typing import Optional
+
+import torch
+from omegaconf.dictconfig import DictConfig
+from pytorch_lightning.plugins.precision.native_amp import NativeMixedPrecisionPlugin
+from pytorch_lightning.trainer.trainer import Trainer
+
+from nemo.collections.nlp.data.language_modeling.megatron.megatron_batch_samplers import (
+    MegatronPretrainingBatchSampler,
+    MegatronPretrainingRandomBatchSampler,
+)
+from nemo.collections.nlp.data.language_modeling.megatron.retro_dataset import build_mock_train_valid_test_datasets
+from nemo.collections.nlp.models.language_modeling.megatron_base_model import MegatronBaseModel
+from nemo.collections.nlp.modules.common.megatron.clip_grads import clip_grad_norm_fp32
+from nemo.collections.nlp.modules.common.megatron.module import Float16Module
+from nemo.collections.nlp.modules.common.megatron.retrieval_token_level_encoder_decoder import (
+    MegatronRetrievalTokenLevelEncoderDecoderModule,
+)
+from nemo.collections.nlp.modules.common.megatron.utils import (
+    ApexGuardDefaults,
+    average_losses_across_data_parallel_group,
+    get_params_for_weight_decay_optimization,
+)
+from nemo.collections.nlp.modules.common.tokenizer_utils import get_nmt_tokenizer
+from nemo.collections.nlp.parts.nlp_overrides import GradScaler
+from nemo.core.optim import MainParamsOptimizerWrapper, prepare_lr_scheduler
+from nemo.utils import AppState, logging
+
+try:
+    from apex.transformer.enums import ModelType
+    from apex.transformer import parallel_state
+
+    HAVE_APEX = True
+except (ImportError, ModuleNotFoundError):
+    ModelType = ApexGuardDefaults()
+    HAVE_APEX = False
+
+
+__all__ = ["MegatronRetrievalModel"]
+
+
+class MegatronRetrievalModel(MegatronBaseModel):
+    """
+    Megatron Retrieval enhanced language model
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer=trainer)
+
+        # build tokenizer (defaults to nemo supported tokenizers)
+        self._build_tokenizer()
+
+        # manipulate vocabulary (e.g., pad vocabulary for better efficiency)
+        self._build_vocab()
+
+        # TODO does not support PP yet
+        self.model = self.model_provider_func(True, True, True, True)
```

Comment:
Can we pass this via kwargs? Hard to understand what True, True, True, True is.

Reply:
fixed